File: csv_converter.py
import pandas as pd
import json

def process_meters_data(df):
    # Clean and rename columns for meters data
    column_mapping = {
        'SL.NO': 'sl_no',
        'LOCATION': 'location',
        'PANEL NAME': 'panel_name',
        'Meter ID\n (Added by env)': 'meter_id',
        'Meter Name': 'meter_name',
        'Meter Serial No  (Added by env)': 'meter_serial',
        'Name of the equipment  meter connected': 'equipment_connected',
        'EM MAKE/MODEL NO': 'model',
        'COMMUNICATION PORT AVAILABILTY': 'comm_port',
        'REMARKS': 'remarks',
        'INTERNAL LOOPING STATUS (TBP/Yes/No)': 'internal_looping_status',
        'NEW METER INSTALLATION REQUIRED (YES/NO)': 'new_meter_required'
    }
    
    df = df.rename(columns=column_mapping)
    df = df.fillna('')
    df = df[df['sl_no'].astype(str).str.strip() != '']
    
    # Convert boolean fields
    df['new_meter_required'] = df['new_meter_required'].apply(
        lambda x: str(x).strip().upper() == 'YES' if x else False
    )
    
    return df.to_dict('records')

def process_panels_data(df):
    # Process panels data which has a different structure
    panels = []
    current_panel = None
    
    for _, row in df.iterrows():
        if not pd.isna(row.iloc[0]) and 'Panel' in str(row.iloc[0]):
            current_panel = str(row.iloc[0]).strip()
        elif current_panel and not pd.isna(row.iloc[0]) and str(row.iloc[0]).strip().isdigit():
            panels.append({
                'panel': current_panel,
                'sl_no': str(row.iloc[0]).strip(),
                'equipment': str(row.iloc[1]).strip(),
                'capacity': str(row.iloc[2]).strip(),
                'power': str(row.iloc[3]).strip(),
                'quantity': str(row.iloc[4]).strip()
            })
    
    return panels

def csv_to_json(csv_file):
    try:
        # Read the main meter data
        meters_df = pd.read_csv(csv_file, nrows=79)
        
        # Read the panel data (starts after many empty rows)
        panels_df = pd.read_csv(csv_file, skiprows=160)
        
        # Process both sections
        meters_data = process_meters_data(meters_df)
        panels_data = process_panels_data(panels_df)
        
        # Create the final data structure
        data = {
            "meters": meters_data,
            "panels": panels_data,
            "summary": {
                "total_meters": len(meters_data),
                "new_meters_required": sum(1 for m in meters_data if m.get('new_meter_required')),
                "comm_ports": {
                    "YES": sum(1 for m in meters_data if m.get('comm_port') == 'YES'),
                    "NO": sum(1 for m in meters_data if m.get('comm_port') == 'NO')
                }
            }
        }
        
        # Calculate model breakdown
        models = {}
        for meter in meters_data:
            model = meter.get('model', '')
            if model:
                models[model] = models.get(model, 0) + 1
        data['summary']['models'] = models
        
        # Write to JSON file
        with open('data/data.json', 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2, ensure_ascii=False)
        
        print("CSV data converted to JSON successfully!")
        print(f"Total meters: {data['summary']['total_meters']}")
        print(f"Total panels: {len(panels_data)}")
        
    except Exception as e:
        print(f"Error converting CSV to JSON: {str(e)}")
        raise

if __name__ == "__main__":
    try:
        csv_to_json('Data.csv')
    except Exception as e:
        print(f"Error: {str(e)}")

File: src/lib/csvConverter.ts
import { promises as fs } from 'fs';
import * as path from 'path';
import Papa from 'papaparse';
import { FaltaData, MeterRecord, PanelRecord, Summary } from '../types';

type MeterStringKeys = Exclude<keyof MeterRecord, 'new_meter_required'> | 'new_meter_required';

const headerMapping: Record<string, MeterStringKeys> = {
  'SL.NO': 'sl_no',
  'LOCATION': 'location',
  'PANEL NAME': 'panel_name',
  'METER ID (ADDED BY ENV)': 'meter_id',
  'METER NAME': 'meter_name',
  'METER SERIAL NO (ADDED BY ENV)': 'meter_serial',
  'NAME OF THE EQUIPMENT METER CONNECTED': 'equipment_connected',
  'EM MAKE/MODEL NO': 'model',
  'COMMUNICATION PORT AVAILABILTY': 'comm_port',
  'REMARKS': 'remarks',
  'INTERNAL LOOPING STATUS (TBP/YES/NO)': 'internal_looping_status',
  'NEW METER INSTALLATION REQUIRED (YES/NO)': 'new_meter_required'
};

const NORMALIZE_WHITESPACE = /\s+/g;

function normalizeHeader(value: string): string {
  return value.replace(/\r?\n/g, ' ')
    .replace(NORMALIZE_WHITESPACE, ' ')
    .trim()
    .toUpperCase();
}

function normalizeText(value: string | undefined | null): string {
  if (!value) return '';
  return String(value).replace(/\r?\n/g, ' ').trim();
}

function normalizeYesNo(value: string): 'YES' | 'NO' | '' {
  const upper = value.toUpperCase();
  if (upper === 'YES') return 'YES';
  if (upper === 'NO') return 'NO';
  return '';
}

function buildSummary(meters: MeterRecord[]): Summary {
  const models: Record<string, number> = {};
  let commYes = 0;
  let commNo = 0;
  let newRequired = 0;

  for (const meter of meters) {
    const model = meter.model.trim();
    if (model) {
      models[model] = (models[model] ?? 0) + 1;
    }

    const commPort = normalizeYesNo(meter.comm_port);
    if (commPort === 'YES') {
      commYes += 1;
    } else if (commPort === 'NO') {
      commNo += 1;
    }

    if (meter.new_meter_required) {
      newRequired += 1;
    }
  }

  return {
    total_meters: meters.length,
    new_meters_required: newRequired,
    comm_ports: { YES: commYes, NO: commNo },
    models
  };
}

function extractMeters(rows: string[][]): { meters: MeterRecord[]; nextIndex: number } {
  if (rows.length === 0) {
    return { meters: [], nextIndex: 0 };
  }

  const headerRow = rows[0].map(normalizeHeader);
  const columnIndex: Partial<Record<MeterStringKeys, number>> = {};

  headerRow.forEach((header, index) => {
    const mappedKey = headerMapping[header];
    if (mappedKey) {
      columnIndex[mappedKey] = index;
    }
  });

  const slIndex = columnIndex.sl_no;
  if (slIndex === undefined) {
    throw new Error('Unable to locate SL.NO column in CSV header.');
  }

  const meters: MeterRecord[] = [];
  let currentIndex = 1;

  while (currentIndex < rows.length) {
    const row = rows[currentIndex] ?? [];
    currentIndex += 1;

    const slRaw = normalizeText(row[slIndex]);

    if (!slRaw && row.every(cell => !normalizeText(cell))) {
      break;
    }

    if (!slRaw) {
      continue;
    }

    const getString = (key: MeterStringKeys): string => {
      const index = columnIndex[key];
      if (index === undefined) {
        return '';
      }
      return normalizeText(row[index]);
    };

    const commPortRaw = getString('comm_port');
    const normalizedComm = normalizeYesNo(commPortRaw) || commPortRaw;

    const meter: MeterRecord = {
      sl_no: slRaw,
      location: getString('location'),
      panel_name: getString('panel_name'),
      meter_id: getString('meter_id'),
      meter_name: getString('meter_name'),
      meter_serial: getString('meter_serial'),
      equipment_connected: getString('equipment_connected'),
      model: getString('model'),
      comm_port: normalizedComm,
      remarks: getString('remarks'),
      internal_looping_status: getString('internal_looping_status'),
      new_meter_required: normalizeYesNo(getString('new_meter_required')) === 'YES'
    };

    meters.push(meter);
  }

  return { meters, nextIndex: currentIndex };
}

function extractPanels(rows: string[][], startIndex: number): PanelRecord[] {
  const panels: PanelRecord[] = [];
  let currentPanel: string | null = null;

  for (let i = startIndex; i < rows.length; i += 1) {
    const row = rows[i];
    const firstCell = normalizeText(row[0]);

    if (!firstCell) {
      continue;
    }

    if (!/^\d+$/.test(firstCell) && /panel/i.test(firstCell)) {
      currentPanel = firstCell;
      continue;
    }

    if (currentPanel && /^\d+$/.test(firstCell)) {
      panels.push({
        panel: currentPanel,
        sl_no: firstCell,
        equipment: normalizeText(row[1]),
        capacity: normalizeText(row[2]),
        power: normalizeText(row[3]),
        quantity: normalizeText(row[4])
      });
    }
  }

  return panels;
}

export async function convertCsvToJson(inputPath: string, outputPath: string): Promise<FaltaData> {
  const csvContent = await fs.readFile(inputPath, 'utf-8');
  const parsed = Papa.parse<string[]>(csvContent, {
    skipEmptyLines: false
  });

  if (parsed.errors.length > 0) {
    const messages = parsed.errors.map(err => err.message).join('; ');
    throw new Error(`Failed to parse CSV file: ${messages}`);
  }

  const rows = parsed.data;
  const { meters, nextIndex } = extractMeters(rows);
  const panels = extractPanels(rows, nextIndex);
  const summary = buildSummary(meters);

  const result: FaltaData = { meters, panels, summary };

  const resolvedOutput = path.resolve(outputPath);
  await fs.mkdir(path.dirname(resolvedOutput), { recursive: true });
  await fs.writeFile(resolvedOutput, JSON.stringify(result, null, 2), 'utf-8');

  return result;
}
